### PR TITLE
Add case conversion to ImplicitDefaultLocale

### LINKED ISSUE #2515: Add case conversion to ImplicitDefaultLocale
##  Behavior of the rule
ImplicitDefaultLocale should not only consider `String.format()`, but also case conversions like `String.toUpperCase()` and `String.toLowerCase()`. 

For example,
```
val s: String = getString()
val t: String? = getString()
s.toUpperCase() // Should be reported
t?.toLowerCase() // Should be reported
s.toLowerCase(Locale.US) // Should not be reported
t?.toUpperCase(Locale.US) // Should not be reported.
```
should be reported.

## Context
The reason for case conversion is the same as formatting function: Default locale may result in unexpected result, so we should recommend developers to pass in Locale.US.

Android has a [lint check](https://cs.android.com/android/platform/superproject/+/master:tools/base/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/LocaleDetector.java?q=LocaleDetector) that checks both case conversion and formatting. Note that it is running the check on the loaded JVM class, which should cover all edge cases like `s.let(it::toUpperCase)`. However, I believe most of the violations will be in its vanilla format which can be caught by detekt, which is also satisfactory to me.

### PR DESCRIPTION
**Pull request by chao2zhang**
Fix #2515

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt [hunk 1]
```diff
@@ -7,26 +7,38 @@ import io.gitlab.arturbosch.detekt.api.Entity
...
+import org.jetbrains.kotlin.psi.KtQualifiedExpression
+import org.jetbrains.kotlin.psi.KtSafeQualifiedExpression
 import org.jetbrains.kotlin.psi.KtStringTemplateExpression
 import org.jetbrains.kotlin.resolve.calls.callUtil.getCalleeExpressionIfAny
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
 
 /**
  * Prefer passing [java.util.Locale] explicitly than using implicit default value when formatting
- * strings.
+ * strings or performing case conversion.
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]:
```suggestion
 * strings or performing a case conversion.
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt [hunk 2]
```diff
@@ -38,6 +50,17 @@ class ImplicitDefaultLocale(config: Config = Config.empty) : Rule(config) {
...
+    }
+
+    override fun visitSafeQualifiedExpression(expression: KtSafeQualifiedExpression) {
+        super.visitSafeQualifiedExpression(expression)
+        check(expression)
+    }
+
+    fun check(expression: KtQualifiedExpression) {
+        // String formatting
         if (expression.receiverExpression.text == "String" &&
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: If this code is moved into a separate method, the comment can be spared.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt [hunk 3]
```diff
@@ -46,14 +69,32 @@ class ImplicitDefaultLocale(config: Config = Config.empty) : Rule(config) {
                 issue, Entity.from(expression),
                 "${expression.text} uses implicitly default locale for string formatting."))
         }
-        super.visitDotQualifiedExpression(expression)
+
+        // Case conversion
+        if (isStringOrNullableString(expression.receiverExpression.getType(bindingContext)) &&
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: If this code is moved into a separate method, the comment can be spared.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt [hunk 4]
```diff
@@ -46,14 +69,32 @@ class ImplicitDefaultLocale(config: Config = Config.empty) : Rule(config) {
...
+private fun KtQualifiedExpression.isCalleeNoArgs(): Boolean {
+    val lastCallExpression = lastChild as? KtCallExpression
+    return lastCallExpression?.valueArguments.isNullOrEmpty()
+}
+
+private fun KtQualifiedExpression.containsStringTemplate(): Boolean {
     val lastCallExpression = lastChild as? KtCallExpression
-    return lastCallExpression?.valueArgumentList
-        ?.arguments
+    return lastCallExpression?.valueArguments
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: Good catch!

Comment (chao2zhang) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: More of it is a shortcut.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: Ack. Let me actually do that.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt [hunk 5]
```diff
@@ -46,14 +69,32 @@ class ImplicitDefaultLocale(config: Config = Config.empty) : Rule(config) {
...
+        // Case conversion
+        if (isStringOrNullableString(expression.receiverExpression.getType(bindingContext)) &&
+            expression.isCalleeCaseConversion() &&
+            expression.isCalleeNoArgs()) {
+            report(CodeSmell(
+                issue,
+                Entity.from(expression),
+                "${expression.text} uses implicitly default locale for string conversion."))
+        }
+        super.visitQualifiedExpression(expression)
```
Comment (chao2zhang) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: This line should be removed

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt (hunk 4 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocale.kt]: That's why I especially like it.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt [hunk 6]
```diff
@@ -25,7 +31,64 @@ class ImplicitDefaultLocaleSpec : Spek({
                 fun x() {
                     String.format(Locale.US, "%d", 1)
                 }"""
-            assertThat(subject.compileAndLint(code)).isEmpty()
+            assertThat(subject.compileAndLintWithContext(wrapper.env, code)).isEmpty()
+        }
+
+        it("report String.toUpperCase() call without explicit locale") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]: To be compliant with the other test case descriptions in this file and in this repository.

```suggestion
        it("reports String.toUpperCase() call without explicit locale") {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt [hunk 7]
```diff
@@ -25,7 +31,64 @@ class ImplicitDefaultLocaleSpec : Spek({
...
+            val code = """
+                fun x() {
+                    val s = "deadbeef"
+                    s.toUpperCase()
+                }"""
+            assertThat(subject.compileAndLintWithContext(wrapper.env, code)).hasSize(1)
+        }
+
+        it("does not Sreport tring.toUpperCase() call with explicit locale") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]:
```suggestion
        it("does not report tring.toUpperCase() call with explicit locale") {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt [hunk 8]
```diff
@@ -25,7 +31,64 @@ class ImplicitDefaultLocaleSpec : Spek({
...
+                import java.util.Locale
+                fun x() {
+                    val s = "deadbeef"
+                    s.toUpperCase(Locale.US)
+                }"""
+            assertThat(subject.compileAndLintWithContext(wrapper.env, code)).isEmpty()
+        }
+
+
+        it("report String.toLowerCase() call without explicit locale") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]:
```suggestion
        it("reports String.toLowerCase() call without explicit locale") {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt (hunk 7 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]:
```suggestion
        it("does not report String.toUpperCase() call with explicit locale") {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt [hunk 9]
```diff
@@ -25,7 +31,64 @@ class ImplicitDefaultLocaleSpec : Spek({
...
+        it("report String?.toUpperCase() call without explicit locale") {
+            val code = """
+                fun x() {
+                    val s: String? = "deadbeef"
+                    s?.toUpperCase()
+                }"""
+            assertThat(subject.compileAndLintWithContext(wrapper.env, code)).hasSize(1)
+        }
+
+        it("report String?.toLowerCase() call without explicit locale") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]:
```suggestion
        it("reports String?.toLowerCase() call without explicit locale") {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt [hunk 10]
```diff
@@ -25,7 +31,64 @@ class ImplicitDefaultLocaleSpec : Spek({
...
+            val code = """
+                import java.util.Locale
+                fun x() {
+                    val s = "deadbeef"
+                    s.toLowerCase(Locale.US)
+                }"""
+            assertThat(subject.compileAndLintWithContext(wrapper.env, code)).isEmpty()
+        }
+
+        it("report String?.toUpperCase() call without explicit locale") {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ImplicitDefaultLocaleSpec.kt]:
```suggestion
        it("reports String?.toUpperCase() call without explicit locale") {
```

**Comment by schalkms** [comment]
The execution of `./gradlew build` results in an updated doc.
https://github.com/arturbosch/detekt/blob/master/docs/pages/documentation/potential-bugs.md

**Comment by chao2zhang** [comment]
The doc should be updated now.

**Comment by schalkms** [comment]
Yep 👍 
That looks good. The test case description could be updated to be compliant with other test descriptions.

**Pull request merged**